### PR TITLE
pyupgrade, Black, isort, Flake8 on CI via pre-commit

### PR DESCRIPTION
**Pull request by hugovk**
Re: https://github.com/urllib3/urllib3/pull/2044#issuecomment-730961021

Run pyupgrade, Black, isort and Flake8 on GitHub Actions via pre-commit.

It uses https://github.com/pre-commit/action for GHA so minimal config.

The `.pre-commit-config.yaml` defines which tools pre-commit will run.

By explicitly setting `py36` in the args for pyupgrade and Black, it finds a couple things missed so far, because when it's implicit sometimes the tool guesses the file might be for an older version.

---

If developers wish, they can run pre-commit locally. I really like it but it's not for everyone, and the CI will catch things anyway. To use locally:

```sh
# One-off installation
python -m pip install pre-commit
pre-commmit install
```
```console
# Work as normal, then commit as normal
# The first time it runs it'll take a minute or two to install some stuff
# The next time it's much quicker
$ git commit -m "pyupgrade, Black, isort, Flake8 on CI via pre-commit"
pyupgrade................................................................Passed
black....................................................................Passed
isort....................................................................Passed
flake8...................................................................Passed
[lint-with-pre-commit ca014f6] pyupgrade, black, isort, flake8 on CI via pre-commit
 4 files changed, 36 insertions(+), 4 deletions(-)
 create mode 100644 .github/workflows/lint.yml
 create mode 100644 .pre-commit-config.yaml
```
```sh
# From time to time, update the pins (or try the alpha preview of https://results.pre-commit.ci)
pre-commit autoupdate
```
```sh
# Not keen? This stops running on commit
pre-commit uninstall
```
```sh
# At any time, even if it's not installed for a repo, 
# you can check all the files in the repo (not just staged)
pre-commit run --all-files
```

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    hooks:
+      - id: black
+        args: ["--target-version", "py36"]
+
+  - repo: https://github.com/PyCQA/isort
+    rev: 5.6.4
+    hooks:
+      - id: isort
+
+  - repo: https://gitlab.com/pycqa/flake8
```
Comment (graingert) [inline .pre-commit-config.yaml:19]:
```suggestion
  - repo: https://github.com/pycqa/isort
    rev: 5.6.4
    hooks:
      - id: isort

  - repo: https://gitlab.com/pycqa/flake8
```

```suggestion
  - repo: https://github.com/PyCQA/isort
    rev: 5.6.4
    hooks:
      - id: isort

  - repo: https://gitlab.com/PyCQA/flake8
```

Comment (hugovk) [reply .pre-commit-config.yaml:19]: I used what the websites actually use:

* GitHub is `PyCQA`: `https://github.com/PyCQA/isort`

* GitLab is `pycqa`: `https://gitlab.com/pycqa/flake8`

Comment (pquentin) [reply .pre-commit-config.yaml:19]: The suggestions are only about capitalizations, not gitlab vs github. Only the GitHub one seems correct, though.

Comment (graingert) [reply .pre-commit-config.yaml:19]: Oh woops

**Review by pquentin — APPROVED** [review]
I like the approach! Not saying that you should be doing that yourself @hugovk, but:

 * We'll probably have to keep `nox -s lint` for mypy where we do some custom stuff but remove the flake8/black/isort
 * The docs need to be updated
 * We need to add flake8-2020 to the flake8 configuration

**Comment by graingert** [comment]
>     * We'll probably have to keep `nox -s lint` for mypy where we do some custom stuff but remove the flake8/black/isort

It's common to run `pre-commit run --all-files` in nox/tox

**Comment by pquentin** [comment]
@graingert It would be redundant with using the pre-commit GitHub Actions though

**Comment by hugovk** [comment]
> * We need to add flake8-2020 to the flake8 configuration

Added!

**Comment by graingert** [comment]
@pquentin you might want to try https://results.pre-commit.ci/login/github

**Comment by sethmlarson** [comment]
@graingert We'll be sticking with GHA for now, relying on fewer services is desirable.

**Comment by hugovk** [comment]
(Resolved conflict)

**Comment by sethmlarson** [comment]
@hugovk Thanks for removing the conflict. We still have another GHA for linting that just calls nox, is it possible to switch to pre-commit for our linting CI completely and then stitch `nox -s lint` to a pre-commit call?

**Comment by hugovk** [comment]
Updated!

**Review by sethmlarson — APPROVED** [review]
LGTM, thank you!

**Pull request merged**